include/ginkgo/core/solver/solver_base.hpp

```diff
@@ -50,6 +51,126 @@ namespace gko {
 namespace solver {
 
 
+/**
+ * Give a initial guess mode about the input of the apply method.
+ */
+enum class initial_guess_mode {
+    /**
+     * the input is zero
+     */
+    zero,
+    /**
+     * the input is right hand side
+     */
+    rhs,
+    /**
+     * the input is provided
+     */
+    provided
+};
+
+
+class MultigridState;
+
+
+/**
+ * ApplyWithInitialGuess provides a way to give the input guess for apply
+ * function. All functionalities have the protected access specifier. It should
+ * only be used internally.
+ */
+class ApplyWithInitialGuess {
+protected:
+    friend class MultigridState;
+
+    /**
+     * Modify the input vector x by the guess
+     *
+     * @param b  the right hand side vectors
+     * @param x  the input vectors
+     * @param guess  the input guess
+     */
+    static void prepare_initial_guess(const LinOp* b, LinOp* x,
```

Comment:
Do you think there might be future cases where some solvers might want to do something specific to prepare the initial guesses and hence maybe this function could be virtual ?

Reply:
Since this is protected and needs to be called by the solver expliclty, I think we can handle this on the `apply_with_initial_guess` side, and theoretically don't need to keep this function.

Reply:
I wouldn't implement it in each time and it is required for default implementation

Reply:
My suggestion is: Move the code to `apply_with_initial_guess`, and maybe provide ValueType templated helpers for implementing this inside apply_dense_impl, but don't use it in solvers that already provide custom support for apply_with_initial_guess, because there you don't need the runtime dispatch.